python/ray/serve/_private/replica_scheduler/replica_scheduler.py

```diff
@@ -47,6 +49,126 @@ class LocalityScope(str, enum.Enum):
     AVAILABILITY_ZONE = "AVAILABILITY_ZONE"
 
 
+class MultiplexScheduleMixin:
+    """Mixin for multiplex scheduling.
+    This mixin is used to schedule requests to replicas that are multiplexed.
+    It adds necessary attributes and methods to keep track of multiplexed
+    model IDs and offer the helpers to apply multiplex scheduling and rank
+    replicas based on multiplexed model IDs.
+    """
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self._multiplexed_model_id_to_replica_ids: DefaultDict[
+            str, Set[ReplicaID]
+        ] = defaultdict(set)
+
+        # When there is no match for a multiplexed model id, we will try to fall back
+        # to all replicas immediately. This set is used to make sure we only fall back
+        # once for concurrent requests for the same model id.
+        # Whenever there is a match, we will remove the model id from this set.
+        self._multiplexed_model_id_fallback_match: Set[str] = set()
+        self._replica_id_set: Set[ReplicaID] = set()
+        self._replicas: Dict[ReplicaID, RunningReplica] = {}
+
+    def update_multiplexed_model_ids_with_replicas(
+        self, replicas: List[RunningReplica]
+    ):
+        """Update the multiplexed model IDs based on the replicas.
+        This is called when the replicas are updated.
+        """
+        new_multiplexed_model_id_to_replica_ids = defaultdict(set)
+
+        for r in replicas:
+            for model_id in r.multiplexed_model_ids:
+                new_multiplexed_model_id_to_replica_ids[model_id].add(r.replica_id)
+
+        self._multiplexed_model_id_to_replica_ids = (
+            new_multiplexed_model_id_to_replica_ids
+        )
+
+    def _get_replica_ids_with_fewest_multiplexed_models(self) -> Set[str]:
+        """Get the set of replicas that have the fewest multiplexed models loaded."""
+        candidates = set()
+        sorted_replicas = sorted(
+            self._replicas.values(), key=lambda x: len(x.multiplexed_model_ids)
+        )
+        least_num_multiplexed_model_ids = math.inf
+        for replica in sorted_replicas:
+            if len(replica.multiplexed_model_ids) <= least_num_multiplexed_model_ids:
+                candidates.add(replica.replica_id)
+                least_num_multiplexed_model_ids = len(replica.multiplexed_model_ids)
+            else:
+                break
+
+        return candidates
+
+    @property
+    def multiplexed_matching_timeout(self) -> float:
+        return random.uniform(
+            RAY_SERVE_MULTIPLEXED_MODEL_ID_MATCHING_TIMEOUT_S,
+            RAY_SERVE_MULTIPLEXED_MODEL_ID_MATCHING_TIMEOUT_S * 2,
+        )
+
+    def apply_multiplex_scheduling(
+        self,
+        pending_request: Optional[PendingRequest] = None,
+    ) -> Set[ReplicaID]:
+        """Apply multiplex scheduling to the pending request.
```

Comment:
can you add `Args` and `Returns` to the docstring, and anything else you think would be helpful? I'm assuming `apply_multiplex_scheduling` will be one of the few examples users have to look at when trying to write their own custom logic, so ideally we document it really well